src/repr/src/scalar.rs

```diff
@@ -3728,6 +3732,118 @@ impl Arbitrary for SqlScalarType {
     }
 }
 
+/// The type of a [`Datum`] as it is represented.
+///
+/// Each variant here corresponds to one or more variants of [`SqlScalarType`].
+///
+/// There is a direct correspondence between `Datum` variants and `ScalarType`
+/// variants.
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, Ord, PartialOrd, Hash, MzReflect)]
+pub enum ReprScalarType {
+    Bool,
+    Int16,
+    Int32,
+    Int64,
+    UInt8, // also includes SqlScalarType::PgLegacyChar
+    UInt16,
+    UInt32, // also includes SqlScalarType::{Oid,RegClass,RegProc,RegType}
+    UInt64,
+    Float32,
+    Float64,
+    Numeric,
+    Date,
+    Time,
+    Timestamp {
+        precision: Option<TimestampPrecision>,
+    },
+    TimestampTz {
+        precision: Option<TimestampPrecision>,
+    },
```

Comment:
I'm wondering whether `precision` could also be omitted. (But it's also ok to consider this out of scope from this PR, if it's not immediately clear.)

Reply:
It's not immediately clear to me. `base_eq` seems to ignore the precision field, but precision matters for some operations, for casting to JSON ([`build_row_schema_field_type`](https://github.com/mgree/materialize/blob/ae822d2354e316926e2930fe02ec00035363b6a6/src/interchange/src/json.rs#L296-L302)), for avro schemas ([`validate_schema_2`](https://github.com/mgree/materialize/blob/ae822d2354e316926e2930fe02ec00035363b6a6/src/interchange/src/avro/schema.rs#L214-L219)), and DB interop with mysql/postgres/sqlserver. It _seems_ a bit like fixed lengths on strings (affects semantics of operations, but not types, morally speaking), but I'm not 100% on it. Maybe leave it out of scope, for now?

Reply:
Ok